File: algorithms/15_ruby/strings/boyer_moore.rb
# Boyer-Moore-Horspool substring search.
def boyer_moore(text, pattern)
  n = text.length
  m = pattern.length
  skip = Array.new(256, m)
  (0...(m - 1)).each { |i| skip[pattern.getbyte(i)] = m - 1 - i }
  i = 0
  while i + m <= n
    j = m - 1
    j -= 1 while j >= 0 && text.getbyte(i + j) == pattern.getbyte(j)
    return i if j.negative?

    i += skip[text.getbyte(i + m - 1)]
  end
  -1
end

raise "search failed" unless boyer_moore("here is a simple example", "example") == 17
raise "search failed" unless boyer_moore("abc", "xyz") == -1

puts "boyer-moore ok"

File: algorithms/16_kotlin/strings/boyer_moore.kt
// Boyer-Moore-Horspool substring search.
fun boyerMoore(text: String, pattern: String): Int {
    val n = text.length
    val m = pattern.length
    val skip = IntArray(256) { m }
    for (i in 0 until m - 1) skip[pattern[i].code] = m - 1 - i
    var i = 0
    while (i + m <= n) {
        var j = m - 1
        while (j >= 0 && text[i + j] == pattern[j]) j--
        if (j < 0) return i
        i += skip[text[i + m - 1].code]
    }
    return -1
}

fun main() {
    check(boyerMoore("here is a simple example", "example") == 17) { "search failed" }
    check(boyerMoore("abc", "xyz") == -1) { "search failed" }
    println("boyer-moore ok")
}
